File: python/Sliding Window/sliding_window_maximum.py
from typing import List
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        """
        Find maximum values in sliding windows of size k using a simple approach.
        
        Approach:
        1. Slide a window of size k through the array
        2. For each window, find the maximum value
        3. Append maximum to result array
        4. Move window one position to the right
        
        Time Complexity: O(n*k) where n is length of nums
        - For each window (n-k+1 windows), we find max in O(k) time
        
        Space Complexity: O(1) excluding output array
        - Only use a few variables regardless of input size
        """
        if not nums or k == 0:
            return []
            
        result = []  # Store maximum values for each window
        
        # Slide window through array
        for i in range(len(nums) - k + 1):
            # Find maximum in current window
            window_max = max(nums[i:i+k])
            result.append(window_max)
            
        return result

class Solution2:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        """
        Find maximum values in sliding windows of size k using a deque for O(n) time.
        Approach:
        - Use a deque to store indices of useful elements for each window.
        - Remove indices that are out of the current window or less than the current value.
        Time Complexity: O(n)
        Space Complexity: O(k)
        """
        if not nums or k == 0:
            return []
        result = []
        dq = deque()  # Store indices
        for i, num in enumerate(nums):
            # Remove indices out of window
            while dq and dq[0] < i - k + 1:
                dq.popleft()
            # Remove indices whose values are less than current num
            while dq and nums[dq[-1]] < num:
                dq.pop()
            dq.append(i)
            # Append max for current window
            if i >= k - 1:
                result.append(nums[dq[0]])
        return result

# -------------------
# Test Cases
# -------------------
def test_sliding_window_maximum():
    s1 = Solution()
    s2 = Solution2()
    # Test case 1: Basic
    nums = [1,3,-1,-3,5,3,6,7]
    k = 3
    expected = [3,3,5,5,6,7]
    assert s1.maxSlidingWindow(nums, k) == expected
    assert s2.maxSlidingWindow(nums, k) == expected
    # Test case 2: k = 1
    nums = [4,2,12,3]
    k = 1
    expected = [4,2,12,3]
    assert s1.maxSlidingWindow(nums, k) == expected
    assert s2.maxSlidingWindow(nums, k) == expected
    # Test case 3: k equals array length
    nums = [9,11]
    k = 2
    expected = [11]
    assert s1.maxSlidingWindow(nums, k) == expected
    assert s2.maxSlidingWindow(nums, k) == expected
    # Test case 4: All elements same
    nums = [2,2,2,2]
    k = 2
    expected = [2,2,2]
    assert s1.maxSlidingWindow(nums, k) == expected
    assert s2.maxSlidingWindow(nums, k) == expected
    # Test case 5: Empty array
    nums = []
    k = 3
    expected = []
    assert s1.maxSlidingWindow(nums, k) == expected
    assert s2.maxSlidingWindow(nums, k) == expected
    print("All test cases passed!")

if __name__ == "__main__":
    test_sliding_window_maximum()

File: C++/Sliding Window/sliding_window_maximum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Simple Approach: Check each window (not efficient)
     * Time: O(n*k) where n is array length
     * Space: O(1) excluding output array
     */
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if (nums.empty() || k == 0) return {};
        
        vector<int> result;
        
        // Slide window through array
        for (int i = 0; i <= nums.size() - k; i++) {
            // Find maximum in current window
            int windowMax = nums[i];
            for (int j = i; j < i + k; j++) {
                windowMax = max(windowMax, nums[j]);
            }
            result.push_back(windowMax);
        }
        
        return result;
    }
};

class Solution2 {
public:
    /*
     * Optimized Approach: Using deque for O(n) time
     * - Maintain decreasing order of values in deque
     * - Store indices instead of values for window bounds check
     * 
     * Time: O(n) - each element pushed/popped at most once
     * Space: O(k) - deque stores at most k elements
     */
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if (nums.empty() || k == 0) return {};
        
        vector<int> result;
        deque<int> dq;  // Store indices
        
        for (int i = 0; i < nums.size(); i++) {
            // Remove indices outside current window
            while (!dq.empty() && dq.front() < i - k + 1) {
                dq.pop_front();
            }
            
            // Remove indices with smaller values
            while (!dq.empty() && nums[dq.back()] < nums[i]) {
                dq.pop_back();
            }
            
            // Add current index
            dq.push_back(i);
            
            // Add maximum for current window
            if (i >= k - 1) {
                result.push_back(nums[dq.front()]);
            }
        }
        
        return result;
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Vector Operations:
 *    vector<int>              // Dynamic array declaration
 *    nums.empty()             // Check if vector empty
 *    nums.size()              // Get vector length
 *    result.push_back(val)    // Add element to vector
 * 
 * 2. Deque Operations:
 *    deque<int>              // Double-ended queue
 *    dq.empty()              // Check if deque empty
 *    dq.front()              // Get first element
 *    dq.back()               // Get last element
 *    dq.pop_front()          // Remove first element
 *    dq.pop_back()           // Remove last element
 *    dq.push_back(val)       // Add element at end
 * 
 * 3. Loop Constructs:
 *    for (int i = 0; i < n; i++)  // Standard for loop
 *    while (!dq.empty())          // While loop with condition
 * 
 * 4. Standard Library:
 *    max(a, b)               // Returns larger value
 * 
 * Note: Solution2 is more efficient because:
 * - Uses O(n) time instead of O(n*k)
 * - Maintains deque of potential maximums
 * - Processes each element only once
 */
